### DISCUSSION TITLE
Docker Build Issues with Trame and Wslink

### DISCUSSION THREAD

**Discussion by ferracioli** [Kitware/trame#534] [Trame/ParaView]
Hi everyone,

Our company team uses a Trame-based image to build a Docker container for our dashboard. Recently, the build process for new versions of this image has stopped working. Below are the important points we've observed. Can anyone advise on how to resolve these issues and get the image build working again?

- Wslink released version 2.0.0 recently
- Since 29/05(after the wslink new release), our build is returning this message even when starting the build from scratch:
Error while finding module specification for 'wslink.launcher' (ModuleNotFoundError: No module named 'wslink')
- We are using this trame image: kitware/trame:1.2-glvnd-runtime-ubuntu20.04-py39, but even in recently updated images from Docker hub it does not work
- It seems the build_server.sh script is not creating the expected folder(empty files and subfolders such logs)
- The suggestion in the link below does not work in our case, since we already are using a proper trame image:
https://discourse.vtk.org/t/docker/11106

**Comment by jourdain** [comment]
So the new image set to use should be `kitware/trame:py3.9-glvnd`.
We stop generating the names you are using 6 months ago. 

You can see the available tags [here](https://hub.docker.com/r/kitware/trame/tags)

**Reply by ferracioli** [reply]
We tried using this new tag, but the build still not worked for us. This is the complete error message shown in the Wslink part:
```Starting server...
/opt/trame/activate_venv.sh: line 1: /deploy/server/venv/bin/activate: No such file or directory
/opt/trame/run.sh: line 20: /deploy/server/launcher.json: No such file or directory
Starting the wslink launcher at
/usr/bin/python: Error while finding module specification for 'wslink.launcher' (ModuleNotFoundError: No module named 'wslink')```

**Reply by jourdain** [reply]
That means the build step failed... Any info on what could have happened?

Also do you [set the client_type in docker](https://github.com/Kitware/trame/blob/master/examples/deploy/docker/VtkRendering/Dockerfile#L5)?

**Reply by jourdain** [reply]
By build step, I mean the trame build.sh execution that create the venv and so on with the setup/ driectory.

**Reply by ferracioli** [reply]
These are the scripts we use in our build, the build_server.sh is the first to run, and the Wslink error message occurs there(the others scripts and the Dockerfile does not run in the tests for this reason):

recreate.sh
```
#!/usr/bin/env bash

echo "Removing server directory"
rm -rf server/

echo "Creating server directory"
./scripts/build_server.sh

echo "Building the image"
./scripts/build_image.sh

echo "Running the image"
./scripts/run_image.sh
```

build_server.sh`
```
#!/usr/bin/env bash
CURRENT_DIR=`dirname "$0"`

# Since Mac doesn't come with realpath by default, let's set the full
# paths using PWD.
pushd .
cd $CURRENT_DIR/..
DEPLOY_DIR=$PWD
cd $CURRENT_DIR/../../..
ROOT_DIR=$PWD
popd

echo $CURRENT_DIR
echo $DEPLOY_DIR
echo $ROOT_DIR

docker run --rm          \
    -e TRAME_BUILD_ONLY=1 \
    -v "$DEPLOY_DIR:/deploy" \
    -v "$ROOT_DIR:/local-app" \
    kitware/trame:py3.9-glvnd
```

build_image.sh(this script uses the dockerfile)
```
#!/usr/bin/env bash
CURRENT_DIR=`dirname "$0"`

echo $CURRENT_DIR

. $CURRENT_DIR/build_server.sh

cd $CURRENT_DIR/..
docker build -t dashboard .
cd -
```

run_image.sh
```
#!/usr/bin/env bash

 
echo "$HOME/dashboard/data"

docker run -it --rm \
    -p 8081:80 \
    --network=postgres-network \
    -v "$HOME:/work" \
    dashboard
```

Dockerfile
```
FROM kitware/trame:py3.9-glvnd

ENV TRAME_CLIENT_TYPE=vue2

# RUN mkdir -p /etc/apache2/ssl
COPY /ssl/trame.crt /etc/ssl/certs/trame.crt
COPY /ssl/trame.key /etc/ssl/private/trame.key
COPY /ssl/001-trame.conf /etc/apache2/sites-available/001-trame.conf

RUN /bin/sh -c a2enmod vhost_alias &&     a2enmod proxy &&     a2enmod proxy_http &&     a2enmod proxy_wstunnel &&     a2enmod rewrite &&     a2enmod headers &&    a2enmod ssl &&     a2dissite 000-default.conf &&     a2ensite 001-trame &&     a2dismod autoindex -f # buildkit

# Create the /deploy directory with trame-user as the owner
RUN install -d -o trame-user -g trame-user /deploy

# Install ParaView-EGL
ARG PV_FILE=ParaView-master-5.11.0-347-g72d21070df-osmesa-MPI-Linux-Python3.9-x86_64.tar.gz
RUN mkdir -p /opt/paraview
COPY paraview/$PV_FILE /opt/paraview
RUN tar -xzvf /opt/paraview/$PV_FILE --strip-components=1 -C /opt/paraview/
ENV TRAME_PARAVIEW=/opt/paraview

# Copy the server directory
COPY --chown=trame-user:trame-user ./server /deploy/server
```

**Reply by jourdain** [reply]
If you add `RUN /opt/trame/entrypoint.sh build` at the bottom, you will be able to skip the `build_server.sh` step.
And if something bad is happening during that step, you might be able to see it?

**Reply by ferracioli** [reply]
I tried including both RUN and ENV parameters suggested in the bottom of the Dockerfile, and ran only the build_image.sh(removing the `. $CURRENT_DIR/build_server.sh`) and run_image.sh scripts. But even skipping the build_server.sh step, the result was the same. This is the complete message shown in the error(the Apache error can be ignored, since it happens since the last working version of the dashboard) :

```
 * Restarting Apache httpd web server apache2                                                                                                       AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.18.0.10. Set the 'ServerName' directive globally to suppress this message
                                                                                                                                             [ OK ]
Starting server...
/opt/trame/activate_venv.sh: line 1: /deploy/server/venv/bin/activate: No such file or directory
/opt/trame/run.sh: line 20: /deploy/server/launcher.json: No such file or directory
Starting the wslink launcher at
/usr/bin/python: Error while finding module specification for 'wslink.launcher' (ModuleNotFoundError: No module named 'wslink')
```

Updated part in the bottom of the Dockerfile:
```
ENV TRAME_CLIENT_TYPE=vue2
RUN /opt/trame/entrypoint.sh build
```

**Reply by jourdain** [reply]
Like I said before `That means the build step failed... `, until you provide the output/info from it, I can't help you.

The change I asked you to make was to move that step into the docker build... Now you need to capture the output into a file and read that output...

**Reply by ferracioli** [reply]
We managed to build the image by skipping the build_server.sh and including `RUN /opt/trame/entrypoint.sh build` at the bottom of the Dockerfile. When running the image, we couldn't access the dashboard, and the error message obtained in the image log was:

```
ParaView is using venv: /deploy/server/venv
Traceback (most recent call last):
  File "/opt/paraview/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/paraview/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/paraview/lib/python3.9/site-packages/paraview/apps/trame.py", line 26, in <module>
    main()
  File "/opt/paraview/lib/python3.9/site-packages/paraview/apps/trame.py", line 22, in main
    module = importlib.import_module(TRAME_APP)
  File "/opt/paraview/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 984, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'dashboard'
root@683eaf8d91f1:/deploy/server/logs/launcher#
```

**Reply by jourdain** [reply]
Your are giving me the output of docker running, not the output of the docker building. 

Maybe you can do `RUN /opt/trame/entrypoint.sh build &> /deploy/server/build-step.txt`

And post the content of that file? (I mean `/deploy/server/build-step.txt` from within the docker)

**Reply by ferracioli** [reply]
Sorry for the delay, when we try to check the error log, the build-step.txt file always stores a message telling the installed libraries. The only exception is the part 
`ERROR: Invalid requirement: '/local-app'
Hint: It looks like a path. File '/local-app' does not exist.`
It seems that this local-app was build during the build_server.sh script, which was skipped(build_server script shown in the bottom of the answer).

Besides that, we always end up with the previous error message, even when including the wslink in the requirements.

```
#!/usr/bin/env bash
CURRENT_DIR=`dirname "$0"`

# Since Mac doesn't come with realpath by default, let's set the full
# paths using PWD.
pushd .
cd $CURRENT_DIR/..
DEPLOY_DIR=$PWD
cd $CURRENT_DIR/../../..
ROOT_DIR=$PWD
popd

echo $CURRENT_DIR
echo $DEPLOY_DIR
echo $ROOT_DIR

 docker run           \
     -e TRAME_BUILD_ONLY=1 \
     -v "$DEPLOY_DIR:/deploy" \
     -v "$ROOT_DIR:/local-app" \
     kitware/trame:py3.9-glvnd
```

**Reply by jourdain** [reply]
It seems pretty obvious that you probably skipped copying your app into `/local-app` within docker.

**Reply by jourdain** [reply]
I mean ideally you should understand what each step is doing but the main goal is to install your project with its dependecies into the virtual environment within docker.

So here, you try to install /local-app into the venv but nothing is there...

**Reply by ferracioli** [reply]
We managed to fix the problem after a lot of searching. Due to some updates in the Trame repository, our build method became deprecated, which was causing the build_server.sh script to stop working properly. We followed the changes made in this commit to resolve the issue: https://github.com/Kitware/trame-cookiecutter/commit/4e8dcd8a6c3be588533c9b8425e7f7e681aed0a6
 
Thank you for your attention and suggestions to help us fix our Dashboard, Jourdain :)
